src/object_access.c

```diff
@@ -0,0 +1,332 @@
+
+		if (strncmp(NameStr(old->relname), NameStr(new->relname), NAMEDATALEN) != 0 ||
+			old->relnamespace != new->relnamespace)
+		{
+			const char *old_ns = get_namespace_name(old->relnamespace);
+			const char *new_ns = get_namespace_name(new->relnamespace);
+
+			rename_hypertable(old_ns, old->relname, new_ns, new->relname);
+		}
+
+		if (old->relowner != new->relowner)
+		{
+			Relation	authrel = heap_open(AuthIdRelationId, AccessShareLock);
+			HeapTuple	authtup = get_object_by_oid(authrel, new->relowner, NULL);
+			Form_pg_authid auth = (Form_pg_authid) GETSTRUCT(authtup);
+
+			process_utility_set_expect_chunk_modification(true);
+			change_hypertable_owner(ht, auth->rolname);
+			process_utility_set_expect_chunk_modification(false);
+
+			heap_close(authrel, AccessShareLock);
+		}
+
+		heap_close(rel, AccessShareLock);
+	}
+	cache_release(hcache);
+}
+
+static void
+alter_column(Oid relid, int16 attnum)
+{
+	Relation	rel = heap_open(AttributeRelationId, AccessShareLock);
+	HeapTuple	oldtup = get_attribute_by_relid_attnum(rel, relid, attnum, NULL);
+	Form_pg_attribute old = (Form_pg_attribute) GETSTRUCT(oldtup);
+
+	if (!old->attisdropped && old->attnum > 0)
+	{
+		Cache	   *hcache = hypertable_cache_pin();
+		Hypertable *ht = hypertable_cache_get_entry(hcache, old->attrelid);
+
+		if (ht != NULL)
+		{
+			HeapTuple	newtup = get_attribute_by_relid_attnum(rel, relid, attnum, SnapshotSelf);
+			Form_pg_attribute new = (Form_pg_attribute) GETSTRUCT(newtup);
+
+			if (strncmp(NameStr(old->attname), NameStr(new->attname), NAMEDATALEN) != 0)
+			{
+				CatalogSecurityContext sec_ctx;
+
+				catalog_become_owner(catalog_get(), &sec_ctx);
+				rename_hypertable_column(ht, old->attname, new->attname);
+				catalog_restore_user(&sec_ctx);
+			}
+			if (old->atttypid != new->atttypid)
+			{
+				CatalogSecurityContext sec_ctx;
+
+				catalog_become_owner(catalog_get(), &sec_ctx);
+				process_utility_set_expect_chunk_modification(true);
+				change_hypertable_column_type(ht, new->attname, new->atttypid);
+				process_utility_set_expect_chunk_modification(false);
+				catalog_restore_user(&sec_ctx);
+			}
+		}
+		cache_release(hcache);
+	}
+
+	heap_close(rel, AccessShareLock);
+}
+
+static void
+create_constraint(Oid constraint_oid)
+{
+	Relation	rel;
+
+	if (!OidIsValid(constraint_oid))
+		return;
+
+	rel = heap_open(ConstraintRelationId, AccessShareLock);
+
+	HeapTuple	newtup = get_object_by_oid(rel, constraint_oid, SnapshotSelf);
+	Form_pg_constraint new = (Form_pg_constraint) GETSTRUCT(newtup);
+
+	if (OidIsValid(new->conrelid))
+	{
+		Cache	   *hcache = hypertable_cache_pin();
+		Hypertable *ht = hypertable_cache_get_entry(hcache, new->conrelid);
+
+		if (ht != NULL)
+		{
+			Assert(strlen(new->conname.data) > 0);
+			process_utility_set_expect_chunk_modification(true);
+			add_hypertable_constraint(ht, new->conname);
+			process_utility_set_expect_chunk_modification(false);
+		}
+		else
+		{
+			if (new->contype == CONSTRAINT_FOREIGN)
```

Comment:
Is the case when the user tries to add Foreign key between two hypertables catched ? This would only catch a FK-constraint on a non-hypertable into a ht right ?